### ISSUE TITLE
Include ReferenceConverter.ctor(Type) with linker by default

### ISSUE THREAD

**Issue by JamesNK** [xamarin/xamarin-macios#3372]
The missing ctor causes this common error - https://github.com/xamarin/xamarin-macios/issues/3144

Neither the end user or Newtonsoft.Json is using that ctor. Something inside the .NET Framework is using reflection to invoke the ctor which causes an error when the linker doesn't include it.

Xamarin should link it by default rather than forcing the user to debug the error, search the web, and then eventually find the issue above and fix it themselves.

**Comment by JamesNK** [comment]
If you are worried about bloating the final package size with unnecessary methods, you could do something fancy like conditionally include `ReferenceConverter.ctor(Type)` when you detect `Newtonsoft.Json` is referenced.

**Comment by dalexsoto** [comment]
We acknowledge this enhancement request, therefore marked it as such.

However, this does not necessarily mean we will take actions on it as we still need to discuss its feasibility internally and make sure it is not conflicting with other features.

**Comment by blmiles** [comment]
@JamesNK you said: "...rather than forcing the user to debug the error, search the web, and then eventually find the issue above and fix it themselves."

Question: what then IS the solution to this while we're awaiting a possible fix?

My app compiles until I call the JsonConvert.SerializeObject(myObject) method.
Then I get: Constructor on type 'System.ComponentModel.ReferenceConverter' not found.

thx

**Comment by JamesNK** [comment]
Having a Linker.xml that includes the constructor is the fix. You can do that now - https://github.com/xamarin/xamarin-macios/issues/3144#issuecomment-361303109

Xamarin should include it by default so developers don't need to jump through these hoops.

**Comment by spouliot** [comment]
@blmiles do you have a test case you can share with us ?

If that `ReferenceConverter` constructor is used by reflection from the BCL/SDK the we should fix this (it's a bug) inside our tools - based on condition where this is needed (not impacting other use cases). However the previous bug report could not be confirmed by the team (only externally).

**Comment by spouliot** [comment]
Since we moved to reference sources it seems there are more type converters then before (in mono). I have not hit `ReferenceConverter` (so a test case would be nice) but we're also missing `NullableConverter` (easy to test).

Still if anyone can share a test case or just the stack trace it would be helpful since we preserve them (the type descriptors) only if the code path can hit the case. I don't think there's a different one but it's easy to confirm from a stack trace and I've been wrong before ;-).

**Comment by blmiles** [comment]
Sorry this has taken a while to reply.

Actually, the error: Constructor on type 'System.ComponentModel.ReferenceConverter' not found... was correct.

Sorry, I cannot post my code but will try with an analogy :)

I have interface IAnimal and IDog : IAnimal and ICat : IAnimal
I have concrete classes Dog : IDog an Cat : ICat
I have a List<IAnimal>() Animals which contains dog n cats.

JsonConvert.SerializeObject(Animals ) was breaking as I was expecting it to serialise the list of Animals, read interfaces...

I worked around it but having lists of Dogs and Cats, loop through the Animals list, add each animal to their respective collection then serialise those.

It meant I did not need to re-factor the whole design using multiple lists as those ONLY get used in serialisation and deserialization thus it also meant adding string fields to the data object DogsJson and CatsJson and not have a AnimalsJson field AND I get to keep my design as-is using the Animals list.

So, I think that error popped up as it should cos I was not using the JsonConvert methods properly.

Thanks
